auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,204 @@
+    auctions: StorageMap<u64, Auction> = StorageMap {},
+    /// Tracking how many auctions have been made till now
+    auction_count: u64 = 0,
+    /// Auction ids of the active auctions by author
+    active_auctions_of_author: StorageMapVec<Identity, u64> = StorageMapVec {},
+    /// Auction ids of all auctions by author
+    auctions_of_author: StorageMapVec<Identity, u64> = StorageMapVec {},
+    /// The auctions which any given bidder has won
+    auctions_won: StorageMapVec<Identity, u64> = StorageMapVec {},
+}
+
+impl DutchAuction for Contract {
+    #[storage(read)]
+    fn price(auction_id: u64) -> u64 {
+        validate_id(auction_id, storage.auction_count);
+        calculate_price(storage.auctions.get(auction_id))
+    }
+
+    #[storage(read, write)]
+    fn bid(auction_id: u64) {
+        // In a Dutch auction the first bid wins
+        validate_id(auction_id, storage.auction_count);
+
+        let mut auction = storage.auctions.get(auction_id);
+
+        require(!auction.ended, TimeError::AuctionAlreadyEnded);
+
+        // Cannot bid before auction starts
+        require(auction.start_time <= height(), TimeError::AuctionNotYetStarted);
+        // Cannot bid after auction ends
+        require(height() <= auction.end_time, TimeError::AuctionAlreadyEnded);
+
+        // Checks for correct asset_id being sent
+        require(msg_asset_id() == auction.asset_id, UserError::WrongAssetSent);
+
+        let price = calculate_price(auction);
+
+        // Checks for high enough amount being sent
+        require(price <= msg_amount(), UserError::BidTooLow);
+
+        // Disallows furthur bids
+        auction.ended = true;
+        auction.winner = Option::Some(sender_indentity());
+        storage.auctions.insert(auction_id, auction);
+
+        // If someone sends more than the current price, refunds the extra amount
+        if msg_amount() > price {
+            let return_amount = msg_amount() - price;
+            transfer_to_identity(return_amount, auction.asset_id, sender_indentity());
+        }
+
+        on_win(auction, price);
+
+        let _ = storage.active_auctions_of_author.pop(auction.author);
+
+        storage.auctions_won.push(sender_indentity(), auction_id);
+
+        log(WinningBidEvent {
+            id: auction_id,
+            winner: sender_indentity(),
+        });
+    }
+
+    #[storage(read, write)]
+    fn create_auction(
+        opening_price: u64,
+        reserve_price: u64,
+        start_time: u64,
+        end_time: u64,
+        beneficiary: Identity,
+        asset: ContractId,
+    ) {
+        require(reserve_price <= opening_price, SetupError::EndPriceCannotBeLargerThanStartPrice);
+        require(height() < end_time, SetupError::AuctionCannotEndInThePast);
+        require(height() <= start_time, SetupError::AuctionCannotStartInThePast);
+        require(start_time < end_time, SetupError::AuctionCannotEndBeforeItStarts);
+
+        let auction = Auction {
```

Comment:
I'd consider adding a `new()` method.

Reply:
I dont see a point, every field needs to be set either way

Reply:
1. There are 2 default values at the end that can be hardcoded and obscured
2. It makes the main file easier to read on a higher level since it's cleaner